### PR TITLE
Hubbard model

### PR DESCRIPTION
**Pull request by bleutooth65**
Major changes: 
- Added the Hubbard model Hamiltonian to generate charge stability diagrams
 - Moved all plotting to the CSD Analysis module
 - Added ability to plot charge stability diagrams with when no capacitances are supplied
 - All charge stability generators find occupation as a function of voltage and not some current or pseudocurrent
Minor changes:
 - Improvement to the way constant interaction model charge stability diagram is made 
 - Bug fixes for the Hough transform code
 - Added plotting wrapper for CSDAnalysis object
 - File name changes

### CODE REVIEW

File: qudipy/chargestability/csd_hubbard.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        Parameters
+        ----------
+        state: state for the creation operator to be acted on
+        position: place where we will try to increase the electron number
+
+        Returns
+        -------
+        The state incremented by 1 in position, or None
+        '''
+        state = copy.copy(state) # to avoid overwrites onto object
```
[the remark this answers is not in the document]
Comment (theycallmesimba) [reply qudipy/chargestability/csd_hubbard.py:344]: Ah yes, I understand now. Thanks.

File: tutorials/Hubbard Charge Stability.ipynb [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    "params = {'U_11':8.36 * 1E-3, 'U_22':9.38 * 1E-3, 'U_12':2.6 * 1E-3, 't_12':0.3 * 1E-3} \n",
+    "\n",
+    "# To generate the capacitance matrix\n",
+    "alpha_1 = (params['U_22'] - params['U_12'])*params['U_11'] / (params['U_11']*params['U_22'] - params['U_12']**2)\n",
+    "alpha_2 = (params['U_11'] - params['U_12'])*params['U_22'] / (params['U_11']*params['U_22'] - params['U_12']**2)\n",
+    "\n",
+    "# Capacitance matrix (this particular formula is derived from the constant interaction model, but just putting numbers in here also works)\n",
+    "cap_matrix = [[alpha_1,1-alpha_1],[1-alpha_2,alpha_2]]\n",
+    "\n",
+    "hub_csd = hub.HubbardCSD(n_sites, n_electrons, cap_matrix, h_mu=True, h_t=True, h_u=True, **params)"
```
[the remark this answers is not in the document]
Comment (theycallmesimba) [reply tutorials/Hubbard Charge Stability.ipynb:114]: I can understand that.

File: tutorials/Hubbard Charge Stability.ipynb [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  {
+   "source": [],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "source": [
+    "## 5. Plot charge stability diagram\n",
+    "\n",
+    "Finally, we can plot the results of the charge stability diagram. In order to do this, we must pass it to the CSDAnalysis object. We can pass an optional parameter corresponding to a list of capacitances, which tells us how to convert from occupation in each dot to a current in order to replicate what an experiment might see. This list must be the same length as the number of sites since in general, each site will have a different coupling to a charge sensor. If no capacitances are passed, a unique color will be given to each occupation. However, this color is meaningless, so further analysis of the object cannot be done unless capacitances are passed."
```
[the remark this answers is not in the document]
Comment (theycallmesimba) [reply tutorials/Hubbard Charge Stability.ipynb:150]: Yes, that makes sense.

File: tutorials/Charge Stability.ipynb [hunk 4]
```diff
@@ -104,32 +103,51 @@
...
    "metadata": {},
    "source": [
-    "## 5. Find the edges of the charge stability diagram\n",
-    "In order for the Hough transform to be used, we need to have a bitmap of where edges are and aren't in the image. So, we need to run some pre-processing to first find the edges, and then transform that map of edges into a bitmap.\n",
+    "## 5. Plot the resulting diagram\n",
+    "\n",
+    "In order to plot this charge stability diagram, we must then pass this object to the the CSDAnalysis object.\n",
     "\n",
-    "In order to find the edges, we can take the derivative of the above charge stability diagram. This is done automatically when csd.generate_csd() is called, and is stored in csd.csd_der. This derivative is the second plot in the cell above."
+    "To simulate real experimental results, we can pass a list of capacitances from each dot to the charge sensor. This list dictates how much a change in the charge of one dot will change the current passing through a charge sensor. This current is dimensionless, but simulates what would be seen in an experiment. \n",
```
[the remark this answers is not in the document]
Comment (bleutooth65) [reply tutorials/Charge Stability.ipynb:118]: You are correct, I have changed it

File: tutorials/Charge Stability.ipynb [hunk 5]
```diff
@@ -140,9 +158,25 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
     "\n",
     "# Adding artificial noise (Will not be needed in practice)\n",
     "np.random.seed(469230) # fix seed for reproducibility\n",
```
[the remark this answers is not in the document]
Comment (bleutooth65) [reply tutorials/Charge Stability.ipynb:182]: Yes and yes, both have been changed

File: tutorials/Hubbard Charge Stability.ipynb [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    "# Number of electrons\n",
+    "n_electrons = 4\n",
+    "# Pamameters\n",
+    "params = {'U_11':8.36 * 1E-3, 'U_22':9.38 * 1E-3, 'U_12':2.6 * 1E-3, 't_12':0.3 * 1E-3} \n",
+    "\n",
+    "# To generate the capacitance matrix\n",
+    "alpha_1 = (params['U_22'] - params['U_12'])*params['U_11'] / (params['U_11']*params['U_22'] - params['U_12']**2)\n",
+    "alpha_2 = (params['U_11'] - params['U_12'])*params['U_22'] / (params['U_11']*params['U_22'] - params['U_12']**2)\n",
+    "\n",
+    "# Capacitance matrix (this particular formula is derived from the constant interaction model, but just putting numbers in here also works)\n",
```
[the remark this answers is not in the document]
Comment (bleutooth65) [reply tutorials/Hubbard Charge Stability.ipynb:111]: Clarified

File: tutorials/Charge Stability.ipynb [hunk 7]
```diff
@@ -195,10 +229,10 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "The spots where the counts in the color bar are the highest places where lines fit the best. We can see that there are 4 peaks that roughly make a rectangle in the upper right hand side of the graph. These correspond to the 4 large occupation transitions seen in the earlier charge stability diagram.\n",
-    "We can convert from these polar coordiantes to the normal slope and intercept using $m = \\frac{-\\text{cos}(\\theta)}{\\text{sin}(\\theta)}$ and $b = \\frac{\\rho}{\\text{sin}(\\theta)}$.\n",
+    "The spots where the counts in the color bar are the highest places where lines fit the best. We can see that there are 4 peaks. These correspond to the 4 large occupation transitions seen in the earlier charge stability diagram.\n",
+    "We can convert from these polar coordinates to the normal slope and intercept using $m = \\frac{-\\text{cos}(\\theta)}{\\text{sin}(\\theta)}$ and $b = \\frac{\\rho}{\\text{sin}(\\theta)}$.\n",
     "\n",
-    "## 8. Peak Finding & Clustering\n",
+    "## 9. Peak Finding & Clustering\n",
```
[the remark this answers is not in the document]
Comment (bleutooth65) [reply tutorials/Charge Stability.ipynb:235]: Done. I also added some comments leading to the docs

**Pull request merged**